File: src/20. Valid Parentheses/20.cpp
/*
 * @Author: Nick Cao
 * @Date: 2021-06-16 18:15:59
 * @LastEditTime: 2021-06-16 18:16:09
 * @LastEditors: Nick Cao
 * @Description: https://leetcode.com/problems/valid-parentheses/
 * @FilePath: /LeetCodeLog/src/20. Valid Parentheses/20.cpp
 */

class Solution {
public:
    bool isValid(string s) {
        int idx = 0;
        vector<char> stack;
        for(char ch : s){
            switch(ch){
                case ')':
                    if(!stack.empty()&&stack.back()=='(') stack.pop_back();
                    else return false;
                    break;
                case ']':
                    if(!stack.empty()&&stack.back()=='[') stack.pop_back();
                    else return false;
                    break;
                case '}':
                    if(!stack.empty()&&stack.back()=='{') stack.pop_back();
                    else return false;
                    break;
                default:
                    stack.push_back(ch);
            }
        }
        return stack.empty();
    }
};

File: src/20. Valid Parentheses/20.js
/*
 * @Author: Nick Cao
 * @Date: 2021-06-16 18:17:48
 * @LastEditTime: 2021-06-16 18:17:49
 * @LastEditors: Nick Cao
 * @Description: https://leetcode.com/problems/valid-parentheses/
 * @FilePath: /LeetCodeLog/src/20. Valid Parentheses/20.js
 */

var isValid = function(s) {   
    const stack = [];
    
    for (let i = 0 ; i < s.length ; i++) {
        let c = s.charAt(i);
        switch(c) {
            case '(': stack.push(')');
                break;
            case '[': stack.push(']');
                break;
            case '{': stack.push('}');
                break;
            default:
                if (c !== stack.pop()) {
                    return false;
                }
        }
    }
    
    return stack.length === 0;
};
